policies/codingstyle.txt

```diff
@@ -289,12 +289,69 @@ options.
 
 In source files, separate functions with one blank line.
 
+                Chapter 6.1: Function Arguments
+
+Most API functions in OpenSSL revolve around a specific type (e.g. SSL_CTX,
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
+
+Where present the TYPE argument should always be first. In constructors or
+object factory type functions (such as a "fetch" function), the key argument
+that identifies the thing be constructed should be placed first (if such an
+argument exists). For example in a "fetch" function the name of the algorithm
+being fetched would be first.
+
+The remaining argument list should be put into order of importance. Deciding the
+relative importance of arguments is necessarily somewhat subjective, but the
+following guidelines should be considered:
+
+1) Any "context" arguments should be first. If multiple contexts are necessary
+   then they should be in order of most general to most specific. In some cases
+   (typically with an OPENSSL_CTX), a context may be NULL to indicate a default
+   context should be used. This does not lessen its importance. Since such a
+   context is always used, even though a default value may be applied, it does
+   not make it an "optional" argument.
+2) Required arguments should be placed after the context.
+3) Optional arguments (i.e. those that are only be used in certain circumstances
+   and therefore may be "NULL" or empty) are placed after required arguments.
+4) Output parameters are placed after optional arguments.
+5) Callbacks are placed after output parameters.
```

Comment:
Your enumeration is optimized for reading it as a the sentences of a single paragraph:

```
The remaining argument list should be put into order of importance. Deciding the
relative importance of arguments is necessarily somewhat subjective, but the
following guidelines should be considered:

1) Any additional "context" arguments should be first. If multiple contexts are
   necessary then they should be in order of most general to most specific. In
   some cases (typically with an OSSL_LIB_CTX), a context may be NULL to indicate
   a default context should be used.
2) Required arguments should be placed after the additional contexts.
3) Optional arguments (i.e. those that are only be used in certain circumstances
   and therefore may be "NULL" or empty) are placed after required arguments.
4) Output parameters are placed after optional arguments.
5) Callbacks are placed after output parameters.
```

As a consequence, it contains frequent repetitions of the formulation

```
    <bar> should be placed after <foo>
```

which is unnecessarily verbose and redundand, since the ordering is implied by the enumeration. Also, the formulations disregard the fact that `<foo>` might not be present. I would suggest to replace it by the following formulation which IMHO is more lucid and easier to read:

```
The remaining argument list should be put into order of importance. Deciding the
relative importance of arguments is necessarily somewhat subjective, but the
following ordering of the arguments should be adhered to:

1) Additional context arguments

   If multiple contexts are necessary then they should be in order of most general
   to most specific. In some cases (typically with an OSSL_LIB_CTX), a context may
   be NULL to indicate a default context should be used.
   
2) Required arguments

3) Optional arguments

   Here optional are arguments are meant to be used in certain circumstances only
   and therefore may be "NULL" or empty.
   
4) Output parameters

5) Function pointers to callbacks functions
```

_(Side Note: I wasn't able extend the the range of this comment box over all lines of the enumeration, because there are too many other comment boxes already. That's why I repeated the original text.)_

Reply:
"following ordering of the arguments should be adhered to:"

This is saying it a little stronger than I had intended. It was my intention to provide some wiggle room for developer/reviewer discretion based on the circumstances of the particular function in question.

Reply:
That was not my intention Is there a way to loosen the formulation without losing the enhanced readability?

Reply:
Maybe "aimed for" instead of "adhered to"?

Reply:
How about a slight tweak to my original wording:

"Deciding the relative importance of arguments is necessarily somewhat subjective, but the following guideline ordering of the arguments should be considered:"

Reply:
Sounds good to me.